qt/widgets/common/src/NotificationService.cpp

```diff
@@ -0,0 +1,68 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidQtWidgets/Common/NotificationService.h"
+#include "MantidKernel/ConfigService.h"
+#include "MantidKernel/Exception.h"
+#include <QApplication>
+
+namespace MantidQt {
+namespace MantidWidgets {
+
+// Key for the "normalize data to bin width" plot option
+const std::string NotificationService::NOTIFICATIONSENABLEDKEY =
+    "Notifications.Enabled";
+
+void NotificationService::showMessage(const QString &title,
+                                      const QString &message, MessageIcon icon,
+                                      int millisecondsTimeoutHint) {
+  if (isEnabled() && isSupportedByOS()) {
+    QSystemTrayIcon sysTrayIcon(qApp);
+    // get the window icon for the app
+    QIcon windowIcon = qApp->windowIcon();
+    // if no icon is set then use the mantid icon
+    if (windowIcon.isNull()) {
+      try {
+        windowIcon = QIcon(":/images/MantidIcon.ico");
+      } catch (std::exception) {
+        // if we cannot use the embedded icon, use a blank one
+        windowIcon = QIcon(QPixmap(32, 32));
+      }
+    }
+    // set this as the window icon otherwise you get a warning on the console
+    sysTrayIcon.setIcon(windowIcon);
+
+    sysTrayIcon.show();
+    sysTrayIcon.showMessage(title, message, icon, millisecondsTimeoutHint);
+
+    sysTrayIcon.hide();
+  }
+}
+
+bool NotificationService::isEnabled() {
+  bool retVal = false;
+  try {
+    retVal = Mantid::Kernel::ConfigService::Instance()
+                 .getValue<bool>(NOTIFICATIONSENABLEDKEY)
+                 .get_value_or(true);
+  } catch (Mantid::Kernel::Exception::FileError) {
+    // The Config Service could not find the properties file
+    // Disable notifications
+    retVal = false;
+  } catch (Poco::ExistsException) {
+    // The Config Service could not find the properties file
+    // Disable notifications
+    retVal = false;
+  }
+  return retVal;
```

Comment:
Would it be better to return the values in the try/catch statements, rather than storing them in a variable?

Reply:
Slightly personal preference here, I dislike having return statements all over the code generally as it can make it harder to read.  However I do accept that it doesn't make much difference in this example.